rfcs/0016-verification/0016-verification.md

```diff
@@ -0,0 +1,117 @@
+---
+Number: "0016"
+Category: Informational
+Status: Draft
+Author: Dingwei Zhang
+Organization: Nervos Foundation
+Created: 2019-03-08
+---
+
+# CKB-Verification
+
+## Abstract
+
+This RFC describes ckb block data verification rules.
+
+Verification rules are split into three part in CKB implementation. `Header Verification`,  `Block Verification` and `Transaction Verification`.
+
+## Specification
+
+1. Header Verification
+
+    1. proof-of-work verification
+
+        Block to be valid it must hash to a value less than the current target. current hash algorithm is Cuckoo Cycle.
+
+    2. number verification
+
+        CKB block header contain block number, it MUST equal parent’s block number incremented by one, the genesis block has a number of zero.
+
+    3. block version verification
+
+        Block version MUST obey consensus rules.
+
+    4. timestamp verification
+
+        Timestamp (Unix’s time in milliseconds) of block timestamp MUST greater than the median timestamp of previous 11 blocks and less than the network-adjusted time + 15 seconds.
+
+    5. difficulty verification
+
+        The canonical difficulty of a block of header is defined as Diff == Max(Diff_last * o_last / o, Diff_last * 2) (o: orphan rate), It MUST greater or equal genesis' difficulty.
+
+1. Block Verification
+
+   1. empty transactions verification
+
+        Block transactions MUST not be empty.
+
+   2. duplicate
+
+        Block transactions MUST be unique.
+
+   3. cellbase
+
+      * position: cellbase MUST be the first transaction in block.
+      * input: cellbase MUST contain one null input, the input's script fill binary with current block number.
+      * output_capacity must less or equal consensus rules specify block reward.
+
+   4. merkle_root
+
+        Verify transaction [merkle-proof][1].
+
+   5. transaction proposal rules
+
+        Commit transaction must be proposed within consensus rules specify proposal window.
+
+   6. uncles
+
+      * uncles hash: The hash MUST match the [CFB][2] serialized uncles.
+      * number: there are at most 2 uncles included in one block.
```

Comment:
Why 2 uncles at most? An inappropriate limit will affect the calculation of orphan rate.

Reply:
any idea on above comment?

Reply:
uncle can be include in a window H(b) - 8 < H(u) < H(b). 
2 uncles at most for one block is not very tight.

Difficulty adjust has an upper limit( double currently ) for one epoch, so we collect uncles as much as we need, not as much as we can.
whether 2 is ok, we may need some test. 
Maybe use a variable for now?